popovers: Migrate emoji popover to Tippy.

shameondev opened the pull request:
## Description:
This PR is part of our ongoing effort to migrate from Bootstrap-based popovers to Tippy.

## Known issues:
- [x] The arrow color is incorrect in the light theme.
- [x] The popover opens on the second click in the compose section.
    _I've spent a lot of time trying to debug why it only opens on the second click, but I haven't had any results. I would really appreciate any help here._
- [x] The popover doesn't open in places other than the compose box.
- [x] The popover used to open in overlay mode for screens with a width less than 768px, but it's currently not working as intended.
---
### Fixes part of #23632.

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -619,71 +619,65 @@ function register_popover_events($popover) {
+            register_popover_events($popover);
+            show_emoji_catalog();
+        },
+        onShown() {
+            change_focus_to_filter();
```
This doesn't seem to be working after you try to open emoji popover for the 2nd time.

shameondev replied:
Fixed, thanks!

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -189,21 +187,22 @@ function refill_section_head_offsets($popover) {
-        $current_message_emoji_popover_elem = undefined;
+        $(emoji_popover_instance.reference).removeClass("reaction_button_visible");
+        $(emoji_popover_instance.reference).parent().removeClass("reaction_button_visible");
+        emoji_popover_instance.destroy();
```
Also this doesn't seem to be getting triggered as I am seeing multiple instances of the emoji popover:
<img width="1207" alt="image" src="https://github.com/zulip/zulip/assets/25124304/38bcf623-babe-4aae-8578-c7966f222984">

amanagr commented:
@shameondev looks great! Left a few comments.

shameondev commented:
Here's the current state of the work:
- Fixed the bugs that @amanagr found.
- Resolved all the problems listed in the PR description.

It's worth take a look at the last commit where I created a mechanism to render popovers as a centered overlay.

timabbott commented on web/src/popover_menus.js:
```diff
@@ -188,11 +188,54 @@ function on_show_prep(instance) {
+        onMount(instance) {
+            $("body").append($("<div>").attr("id", "popover-overlay"));
+            popover_props.onMount(instance);
+        },
```
How exactly does this block work? I don't see what attaches the popover to `#popover-overlay`

shameondev replied:
`#popover-overlay` is an empty container used for background styling. It does nothing else and is attached to the `body`.

Our popovers are also attached to the `body` element, so they are siblings with `#popover-overlay`.

shameondev replied:
Copying and pasting the `get_props_for_popover_centering` and `show_popover_menu` in #26375 makes me think that creating `get_props_for_popover_centering` should be in a separate preparatory commit that we can merge safely.

timabbott replied:
OK -- should this be called `popover-overlay-background` then?

shameondev replied:
Yeah, I renamed that and also made a separate commit (f8c57ebe) to introduce `get_props_for_popover_centering`.

timabbott commented on web/styles/popovers.css:
```diff
@@ -752,6 +752,11 @@ ul {
+    /* Fixes incorrect emoji popover rendering on small screens in edit mode.  */
+    [data-reference-hidden] {
+        display: inherit;
+    }
```
Can you explain the interaction this addresses in more detail?  I'd like this comment to explain the problem enough details (like what CSS this overrides) that one wouldn't need to re-debug to figure out what this does and if/when it could be removed if we make other changes.

shameondev replied:
Yeah, sure. This definitely needs more details to be written. Does something like this sound good?:
```javascript
/* By default, the `[data-reference-hidden]` attribute has a `display: none` value.
In this case, the emoji popover doesn't show in edit mode when we try to render it as a
centered overlay, because Tippy sets this attribute to the popover. Since we don't need
to hide this emoji popover when the reference is hidden, this solution is chosen. */

```

shameondev replied:
In general, I have a feeling that with this code, we're solving the effect and not the cause and there may be other solutions I haven't found yet.

@amanagr, maybe you could look at this? You have experience with `[data-reference-hidden]` and the problems around that.

amanagr replied:
This seems like the correct solution for this.

timabbott replied:
That comment is better but to be honest I still don't understand it. The type of comment I'd want is (I'm guessing base done the name what to say -- no idea if this is correct):

"[data-reference-hidden] is set by Tippy when the element the popover is based on is hidden. By default, it has `display: none`, in order to automatically have the popover disappear if a rerender causes the element that this was based on to no longer exists.

For the emoji picker when editing a previously sent message message, this is problematic for mobile size windows because <explanation of the situation where it's bad>, so we disable this Tippy feature for that popover. We avoid having the popover remain visible if when the element element the popover was pointed to was moved by <whatever this PR does to avoid that>"

(Writing this down makes me worried that this code could introduce a bug where if someone deletes a message you've opened the reaction picker for, we'll be left with an orphaned emoji picker)

shameondev replied:
My suspicion about this solution was confirmed. I debugged this with a clear head and realized that for mobiles, when using the newly created centered overlay API, we set the reference with `top: 0; left: 0` - which is out of the viewport. Tippy considered this as a hidden reference and hid the popover.

I removed this CSS fix and added the solution in this commit: [6a5dfd75](https://github.com/zulip/zulip/pull/26354/commits/6a5dfd75bab8d6e473ddc1ca9e8b4ddd615acb38). I'm not sure if the implementation is good enough, so I put it in a separate commit, which we could squash later.

timabbott replied:
OK, this makes sense, squashed!

timabbott commented on web/src/emoji_picker.js:
```diff
@@ -620,9 +620,6 @@ function register_popover_events($popover) {
 export function build_emoji_popover($elt, id) {
-    const template_args = {
-        class: "emoji-info-popover",
-    };
```
For this sort of removal commit, a proper commit message should explain why it's correct -- e.g. in this case a sentence "Only one value was ever used."

I know it seems obvious, but if you don't say that, a careful reviewer needs to check this themselves.

shameondev replied:
Got it! Is there anything I should do now, or just to think about this in the future?

timabbott replied:
Yes, next time just extend the commit message when you're touching the PR to answer the question.

(It's very helpful to have this sort of information in the commit message, and not just lost somewhere in the PR review thread, because then it's handy if months from now we discover the change breaks something)

timabbott commented on web/styles/reactions.css:
```diff
@@ -147,7 +147,6 @@
     }
 
     .emoji-popover {
-        display: block;
```
Similarly with this one -- if you say why it's unnecessary, that is a huge help to a reviewer.

shameondev replied:
I understood the concept, thanks!

This is just a cleanup that should act as a preparatory commit. The `.emoji-popover` div element has a `block` value by default.

timabbott replied:
Added to the commit message.

shameondev replied:
Thanks!

timabbott commented on web/src/emoji_picker.js:
```diff
@@ -673,20 +681,7 @@ export function toggle_emoji_popover(element, id, coming_from_actions_popover) {
-    }
-
-    if (user_status_ui.user_status_picker_open()) {
-        build_emoji_popover($elt, id);
```
It seems like this commit deletes the special logic for the `user_status_ui.user_status_picker_open` case without explanation for why that might not be correct. I suspect this is wrong -- I don't think `reaction_button_visible` is a class that's intended to be toggled on that element?

In general this commit would have been much easier to review if you broke it up into pieces -- the parts that are just moving code and easily verified as equivalent with a few words of text explaining why that's the case first, and then anything like that this that might be a functional change with a clear explanation in its commit message for why it's correct.

timabbott replied:
Testing, I find that the user status picker seems to mostly work but if you click the button to pop it 3 times in a row, it doesn't toggle properly -- the second click doesn't do anything visible, so it's definitely bugged.

shameondev replied:
> Testing, I find that the user status picker seems to mostly work but if you click the button to pop it 3-4 times in a row, it doesn't toggle properly -- the first click opens, second closes, and third just appears the tooltip for the element, so it's definitely bugged somehow.

Fixed this in [2fb09b11eca93f3c04d37329139ac0be0176aaf2](https://github.com/zulip/zulip/pull/26354/commits/2fb09b11eca93f3c04d37329139ac0be0176aaf2)

shameondev replied:
> In general this commit would have been much easier to review if you broke it up into pieces -- the parts that are just moving code and easily verified as equivalent with a few words of text explaining why that's the case first, and then anything like that this that might be a functional change with a clear explanation in its commit message for why it's correct.

Got it!

shameondev replied:
> It seems like this commit deletes the special logic for the user_status_ui.user_status_picker_open case without explanation for why that might not be correct.

Yeah, it should be a separate commit to show the reviewer that it's a functional change.

Now we add the "reaction_button_visible" class to every reference element of the emoji popover. There's no need to handle the user status picker case since the reference element should always be visible. That's why we're reducing some complexity, even though it means adding a class to the user status picker reference.

timabbott replied:
OK. Please add the explanation you wrote here to the relevant commit message.

shameondev replied:
Added!

timabbott commented on web/src/popover_menus.js:
```diff
@@ -188,6 +188,23 @@ function on_show_prep(instance) {
+        showOnCreate: true,
+        ...popover_props,
+    });
+}
```
Should `register_popover_menu` be using this as a subroutine? It seems to mostly duplicate code from there.

Also in this commit you're missing a blank line after this function.

shameondev replied:
Thanks! I added a follow-up commit with some refactoring.

timabbott commented on web/src/popover_menus.js:
```diff
@@ -793,11 +793,7 @@ export function initialize() {
-                );
+                emoji_picker.show_emoji_popover(instance.reference.parentElement, message_id);
                 instance.hide();
             });
```
So this change in naming makes me worry that clicking the emoji popover opening button twice will not hide it.

timabbott replied:
That does seem to work fine -- it's just a built-in feature of the new library. So I think we need to:
* Call the function that you're using here `emoji_picker.toggle_emoji_popover` -- basically it looks like you merged those two functions but kept the wrong name.
* Also name your new helper function `popover_menus.show_popover_menu` instead `popover_menus.toggle_popover_menu` so that it's very clear what its expected behavior is.

shameondev replied:
Yeah, these points are reasonable. I changed the names of the functions to the ones you proposed. Thanks!

timabbott commented on web/src/emoji_picker.js:
```diff
@@ -189,21 +188,22 @@ function refill_section_head_offsets($popover) {
 export function reactions_popped() {
-    return $current_message_emoji_popover_elem !== undefined;
+    return emoji_popover_instance;
 }
```
So am I correct that this is actually not just for the reactions special case? 

First, this function should return a bool, not the instance -- that seems unintentionally changed here.

And then let's add a new last commit to rename this function to `emoji_picker.is_open`, since that's a much clearer description of what it checks.

shameondev replied:
> So am I correct that this is actually not just for the reactions special case?

Reactions is just what we call the emoji popover; it's not a special case.

> First, this function should return a bool, not the instance -- that seems unintentionally changed here.

Is it okay to return something like `Boolean(emoji_popover_instance)`? I'm not sure that a comparison with `undefined` might be 100% safe (if someone sets it to `null` somewhere, for example).

> And then let's add a new last commit to rename this function to emoji_picker.is_open, since that's a much clearer description of what it checks.

Sure!

timabbott replied:
> Is it okay to return something like Boolean(emoji_popover_instance)? I'm not sure that a comparison with undefined might be 100% safe (if someone sets it to null somewhere, for example).

I think `=== undefined` is probably better, or `=== null` if that's the convention being used in this file, which might be better. But I don't see why we'd be worrying about writing the code to expect both possibilities when we're very much writing all the logic that sets this variable right here in this file.

timabbott replied:
I left this as `Boolean()` but should flag you put the rename of the function itself in the main Bootstrap commit -- which meant that commit didn't run.

shameondev replied:
Hmm, I thought about making this renaming a separate commit. I might have missed something while rebasing.

timabbott commented on web/src/emoji_picker.js:
```diff
@@ -189,21 +188,22 @@ function refill_section_head_offsets($popover) {
 
 export function hide_emoji_popover() {
-    $(".has_popover").removeClass("has_popover has_emoji_popover");
+    current_message_id = null;
```
Why is this line outside the `reactions_popped` check below?

I think maybe it would make sense to refactor this function to just start with:
```
if {!is_open()) {
    return;
}
```

since I believe all of the logic here is clearly not intended to run if the popover wasn't open in the first place.

Probably can be a follow-up commit.

shameondev replied:
That refactoring helped to catch a bug in the user status picker. Made this as a follow-up commit.

timabbott commented on web/src/emoji_picker.js:
```diff
@@ -619,79 +620,66 @@ function register_popover_events($popover) {
+            emoji_popover_instance = instance;
+            const $popover = $(instance.popper);
+            $popover.addClass("emoji-popover-root");
+            instance.setContent(ui_util.parse_html(template));
```
Do we need to be using this function? Can you explain why it's necessary and safe? 

```
/**                                                                                                     
 * Parse HTML and return a DocumentFragment.                                                            
 *                                                                                                      
 * Like any consumer of HTML, this function must only be given input                                    
 * from trusted producers of safe HTML, such as auto-escaping                                           
 * templates; violating this expectation will introduce bugs that are                                   
 * likely to be security vulnerabilities.                                                               
 */                                                                                                     
export function parse_html(html: string): DocumentFragment {                                            
    const template = document.createElement("template");                                                
    template.innerHTML = html;                                                                          
    return template.content;                                                                            
}                                                                                                       
```

timabbott replied:
I guess I'd feel a lot more comfortable if you just had `render_emoji_popover()` inline here, which seems like the only value `template` ever has? It's a very weird design to have that parameter.

shameondev replied:
The `setContent` method expects an HTML element, not a string in this case. That's why we use the `parse_html` function around the codebase.

shameondev replied:
> I guess I'd feel a lot more comfortable if you just had render_emoji_popover() inline here, which seems like the only value template ever has? It's a very weird design to have that parameter.

Yeah, I agree. I changed it to an inline value.

timabbott commented on web/src/emoji_picker.js:
```diff
@@ -746,7 +734,9 @@ export function register_click_handlers() {
         const message_id = rows.get_message_id(this);
-        toggle_emoji_popover(this, message_id);
+        const target =
+            $(e.target).closest(".emoji_message_control_button_container")[0] ?? e.target;
```
This is pretty weird logic and I imagine is important for this function to work with the several different places we do this popover -- it deserves a comment explaining that.

shameondev replied:
For context, we have this markdown around this exact popover.
```html
<div class="reaction_button" data-tippy-content="{{t 'Add emoji reaction' }}" aria-label="{{t 'Add emoji reaction' }} (:)">
    <div class="emoji-message-control-button-container">
        <i class="zulip-icon zulip-icon-smile" role="button" aria-haspopup="true" tabindex="0" aria-label="{{t 'Add emoji reaction' }} (:)"></i>
        <div class="message_reaction_count">+</div>
    </div>
</div>
```

shameondev replied:
The `.reaction_button` container has padding, so the `.emoji-message-control-button-container` doesn't fully expand inside the `.reaction_button`. If you click inside this "dead zone," then `$(e.target).closest(".emoji-message-control-button-container")[0]` will be undefined. So we use `e.target` as a fallback.

I tried to wrap `.reaction_button` with `.emoji-message-control-button-container`, but it broke the logic. I tried to use CSS to solve this, but it didn't work. I found this fallback value to be the easiest and safest solution.

![image](https://github.com/zulip/zulip/assets/53193850/1eda3a15-60a6-4d9b-a9e3-ef333a22e453)

shameondev replied:
We can add negative margins to this particular `.emoji-message-control-button-container` element (maybe by creating a new class for this) and add padding. We can also change the event listener target to `.emoji-message-control-button-container` instead of `.reaction_button`. This might help to remove this odd logic, but it will add some strange CSS.


```diff
-     $("#main_div").on("click", ".reaction_button", function (e) {
+     $("#main_div").on("click", ".emoji-message-control-button-container", function (e) {
```

```diff
- const target = $(e.target).closest(".emoji-message-control-button-container")[0] ?? e.target;
+ const target = e.currentTarget;
```

timabbott replied:
OK. I'm not happy with this detail but also am tired of re-reviewing this PR; perhaps @karlstolley can try to figure out a follow-up cleanup here.

shameondev replied:
Yeah, I'm not happy with this either. I would appreciate any advice from @karlstolley or anyone else.

timabbott commented on web/styles/reactions.css:
```diff
@@ -73,10 +73,13 @@
 
-    .reaction_button {
+    .reaction_button,
+    .emoji_message_control_button_container {
```
Should this new class name use dashes?

shameondev replied:
Yeah, sure! Changed the underscores to dashes.

timabbott commented on web/styles/zulip.css:
```diff
@@ -636,6 +636,16 @@ body.has-overlay-scrollbar {
     }
 }
 
+#popover-overlay {
```
Let's add a brief comment at this:

/* Styles for popovers rendered as an overlay at the center of the screen, primarily used in mobile sizes. */

shameondev replied:
Added, thanks!

timabbott commented on web/styles/popovers.css:
```diff
@@ -738,11 +738,33 @@ ul {
+    }
+
+    .tippy-box[data-placement="bottom"] .tippy-arrow::before {
+        border-bottom-color: hsl(0deg 0% 93%);
```
The "we consider" comment is confusing to me -- what happens for left/right placement? If the claim is that this will never have left/right placement, we should say that, not just that this code only considers the other possibilities. 

And what is the purpose of having this custom CSS for just this popover, and not others? Generally we want them to have similar styling, so some explanation of what this is for would be appropriate.

shameondev replied:
This comment is wrong. We don't limit where this popover can be placed. The main point here is that the emoji popover has different background color for the header and footer, so we change the arrow to match these colors.

shameondev replied:
I've fixed the comment.

timabbott reviewed (commented):
@shameondev thanks for the work on this. I posted a lot of comments on readability/reviewability, as well as a few classes of bugs that I suspect may be present after reading the code.
.

timabbott commented:
@shameondev is this ready for a next review yet?

shameondev commented:
@timabbott, It wasn't finished at that moment, but it's finished now.

I made the changes you proposed and fixed the issues.

One thing I left unchanged is [this thread](https://github.com/zulip/zulip/pull/26354#discussion_r1298939704).

shameondev commented:
Just resolved the merge conflicts with the last push. It's still ready for review.

timabbott commented:
I reviewed this again, making a few changes and pushing them back here. I noted the following issues that I think are enough that I can't merge this and leave the follow-ups for a future PR:

* This PR has a broken experience if one tries to open the emoji picker when zoomed to ~150% and doing this on a message in the middle of the screen -- the picker goes off the top of the screen because there is no space either above or below it for the full popover. ![image](https://github.com/zulip/zulip/assets/2746074/da59293a-f8b4-4cfb-9352-8e1b131080ee). I'm not sure what the right fix is for this.
* If I click the emoji picker button in the reactions row in a message repeatedly, I get these focus outlines:
![image](https://github.com/zulip/zulip/assets/2746074/fed54269-b342-4e45-8953-ac0beeb991b6)
* Playing around with just clicking on the emoji picker buttons in a message many times very quickly, I was able to reproduce it getting into a state where the emoji picker for that message would no longer open (E.g. hitting `:` did nothing with the message selected). 
* I managed once managed to also pick the `:octopus:` emoji in the status picker and have that end up being applied to a message in the background, rather than set as a new status for the user. I couldn't figure out how to reproduce this, but this certainly is the kind of thing that should be impossible.
* A good follow-up to this might be renaming `emoji-info-popover` to `emoji-picker-popover`, since it's not about information about emoji at all -- it's an emoji picker.
* Similarly `reaction_button_visible` should be renamed to `active-emoji-picker-reference` to clearly document what it's for.

timabbott commented on web/src/popover_menus.js:
```diff
@@ -276,16 +275,7 @@ export function register_popover_menu(target, popover_props) {
-            showOnCreate: true,
-            ...popover_props,
-        });
+        toggle_popover_menu(instance, e.currentTarget, popover_props);
```
This commit would be better squashed into the earlier commit that introduces this function, so we don't have the messy intermediate version. Then it could have a "popover_menus: Extract toggle_popover_menu." commit message and clearly just extract a new function. And then it could be early in the branch and merged to unblock other PRs to use this without an intermediate state with duplicate code.

I also don't understand why it needs both the `target` parameter and te the `instance` parameter when the `instance` always equals `target._tippy`; will future callers do something different?

shameondev replied:
I agree with both, so:
- I squashed the changes into a55d9020
- I removed the redundant instance parameter, so `toggle_popover_menu` calculates `instance` from `target`.

timabbott replied:
Great!

timabbott commented:
I merged a couple prep commits ending with 89e7c79e838b924952191106c73e642e446dc12d.

shameondev commented:
> This PR has a broken experience if one tries to open the emoji picker when zoomed to ~150% and doing this on a message in the middle of the screen -- the picker goes off the top of the screen because there is no space either above or below it for the full popover.

This isn't a regression caused by the library change. For follow-up work, we can look into the options with Popper to prevent overflowing, or try to use centered placement with the new API if we find a way to detect that overflow has happened.

> If I click the emoji picker button in the reactions row in a message repeatedly, I get these focus outlines

This isn't a regression either. I might fix this in one of the prep commits so we can merge it independently.

shameondev commented:
> Playing around with just clicking on the emoji picker buttons in a message many times very quickly, I was able to reproduce it getting into a state where the emoji picker for that message would no longer open (E.g. hitting : did nothing with the message selected).

Even hitting the message actions (:) did nothing, or just clicking on the smiley icon that opens an emoji picker?


> I managed once managed to also pick the :octopus: emoji in the status picker and have that end up being applied to a message in the background, rather than set as a new status for the user. I couldn't figure out how to reproduce this, but this certainly is the kind of thing that should be impossible.

One thing comes to mind - we might not have cleared the `current_message_id` correctly, so the emoji applied to the message rather than the user status.

---

I'm not sure how we should proceed with these issues. Should we take some more time to debug or merge to have an opportunity to get more context from folks on CZO?

shameondev commented:
> If I click the emoji picker button in the reactions row in a message repeatedly, I get these focus outlines:

Addressed in the follow-up commit (45477d5f).

timabbott commented on web/src/popover_menus.js:
```diff
@@ -190,6 +190,24 @@ function on_show_prep(instance) {
 
+// This function is created to instantly toggle the popover menu instead of
+// registering events on `body`. This is useful for cases when we,
+// for example, trigger the popover by a keyboard shortcut.
```
Nit: "This function is created" is something you never want to put in a code comment, because while it's a very reasonable way to talk about a change being made, someone who is reading the code as it exists in the codebase later just wants to read an explanation of the code itself, not its change history, which is invisible to them.

timabbott replied:
I'll rewrite this comment since I think it's the only blocker for merging this commit.

shameondev replied:
Thanks, Tim!

timabbott commented:
I merged 3 more prep commits as the series ending with 53e081ae293466dece742429b76f06e7fe7a58d3; in particular, that should I think unblock other PRs using the new API, such that this PR just contains the emoji picker stuff.

timabbott commented:
> Even hitting the message actions (:) did nothing, or just clicking on the smiley icon that opens an emoji picker?

Correct, I'd say it was actually more reliably broken for the `:` keyboard shortcut after doing this than clicking the icon.

shameondev commented:
In this update:
- Fixed the emoji popover placement to match what's in CZO.
- Added a fallback for when there's limited space above or below. Now the popover will show on the left or right.
- Cleaned up the smiley icon's outline and some aria attributes.

Still need to fix:
- Issues with the `:` shortcut.

alya commented:
Looks good to me in manual testing!

timabbott commented:
You can squash the fixup commits.

I still see a bug that in Chrome in narrow windows, if you click this emoji button, the background fades but the picker fails to appear. The same issue reproduces for me in the "Set a status" UI as well.

![image](https://github.com/zulip/zulip/assets/2746074/46e28c0a-e7dd-4986-8742-6afe5873b488)

shameondev commented:
@timabbott, does the popover appear in the DOM? If it does, but with the `data-hidden-reference` attribute, this commit (a673f7f) might be a solution for this. I encountered this behavior while I was migrating the `user_card_popover` module.

I tested it in Safari, Chrome, and Firefox browsers, and it worked in all of them. Do you want me to cherry-pick this commit here for you to test?

Also, squashed the fixup commits.

timabbott commented:
I grabbed the 3 commits from #26613. I also added the following to address the fact that we were falling back to `left` immediately when using the mouse buttons (where the default placement is configured to be "bottom"):

``` diff
diff --git a/web/src/emoji_picker.js b/web/src/emoji_picker.js
index a3c3260262..25b0ad1bbd 100644
--- a/web/src/emoji_picker.js
+++ b/web/src/emoji_picker.js
@@ -631,7 +631,10 @@ function get_default_emoji_popover_options() {
                     // the opposite bottom or left placement will be used.
                     name: "flip",
                     options: {
-                        fallbackPlacements: ["bottom", "left", "right"],
+                        // We list both bottom and top here, because
+                        // some callers override the default
+                        // placement.
+                        fallbackPlacements: ["bottom", "top", "left", "right"],
                     },
                 },
             ],

```

timabbott commented:
This that detail fixed, this looks good, marked to merge once CI passes.

shameondev commented:
Thanks, Tim!
